daft/dataframe/dataframe.py

```diff
@@ -177,6 +179,54 @@ def _broadcast_query_plan(self, plan_time_start: datetime, plan_time_end: dateti
             plan_time_end=plan_time_end,
         )
 
+    @DataframePublicAPI
+    def pipe(
+        self,
+        func: Callable[..., T],
```

Comment:
we can use `typing_extensions` for <= 3.10. 

something like
```py
    if sys.version_info >= (3, 10):
        from typing import Concatenate, ParamSpec
    else:
        from typing_extensions import Concatenate, ParamSpec

T = TypeVar("T")
P = ParamSpec("P")


class DataFrame:
    def pipe(
        self,
        function: Callable[Concatenate[DataFrame, P], T],
        *args: P.args,
        **kwargs: P.kwargs,
    ) -> T:
```

Reply:
Nice, I've updated to match `daft/api_annotations.py`